### PR TITLE
[Android]: Database locale issues

### PR DESCRIPTION
**Pull request by Krizzu**
Summary:
---------

Trying to address: #422 #178 

Those issues we've seen are related to locale changing and database connection not being closed properly. I could not reproduce it myself, so those are just **preventive measures**. Can't tell if this would help for sure.

The fix is to close db connection when activity is destroyed - it'll be reestablished on next startup (in case of locale change, activity is recreated).

~~Another one is to disable [localized collators for db](https://developer.android.com/reference/android/database/sqlite/SQLiteDatabase#NO_LOCALIZED_COLLATORS), which is the most suggested fix for `Failed to change locale for db` error.
I've moved this fix behind a flag that you need to set in `gradle.properties` in order to work.~~

This has been dropped due to possibility of unexpected behavior. [More info about being consistent with locale collation in android docs.](https://developer.android.com/reference/kotlin/android/database/sqlite/SQLiteDatabase#no_localized_collators)

---

If anyone is able to constantly reproduce the issue, please do let me know. You can try out this branch before it's released by pointing your `package.json` dependency to:

```json
"@react-native-community/async-storage": "git+https://github.com/react-native-community/async-storage.git#fix/ensure-db-closed"
```

### CODE REVIEW

**Comment by kevinranks** [comment]
Would love to see this reviewed & merged soon if possible. We have many frustrated users that are receiving the issue in #422.

File: android/src/main/java/com/reactnativecommunity/asyncstorage/ReactDatabaseSupplier.java [hunk 1]
```diff
@@ -85,7 +86,13 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
         if (tries > 0) {
           deleteDatabase();
         }
-        mDb = getWritableDatabase();
+
+        if(disableLocalizedCollators) {
```
Comment (tido64) [inline android/src/main/java/com/reactnativecommunity/asyncstorage/ReactDatabaseSupplier.java:90]: Do we really need the class member?
```suggestion
        if (BuildConfig.AsyncStorage_noLocalizedCollators) {
```

File: android/build.gradle [hunk 2]
```diff
@@ -33,6 +33,10 @@ def useDedicatedExecutor = rootProject.hasProperty('AsyncStorage_dedicatedExecut
         ? rootProject.properties['AsyncStorage_dedicatedExecutor']
         : false
 
+def disableLocalizedCollators = rootProject.hasProperty('AsyncStorage_noLocalizedCollators')
+        ? rootProject.properties['AsyncStorage_noLocalizedCollators']
+        : false
```
Comment (tido64) [inline android/build.gradle:38]: Nit: We don't really need the ternary but I see we've done the same above...
```suggestion
def disableLocalizedCollators = rootProject.hasProperty('AsyncStorage_noLocalizedCollators')
        && rootProject.properties['AsyncStorage_noLocalizedCollators']
```

**Review by tido64 — APPROVED** [review]

File: android/src/main/java/com/reactnativecommunity/asyncstorage/ReactDatabaseSupplier.java (hunk 1 above)
Comment (Krizzu) [reply android/src/main/java/com/reactnativecommunity/asyncstorage/ReactDatabaseSupplier.java:90]: True, nice catch

File: android/build.gradle (hunk 2 above)
Comment (Krizzu) [reply android/build.gradle:38]: Yes, I kept it for consistency

**Comment by Krizzu** [comment]
Thanks @tido64 for having a look 🙏 

According to [Android docs, one needs to be consistent with what collation rules were used when creating the database](https://developer.android.com/reference/kotlin/android/database/sqlite/SQLiteDatabase#no_localized_collators). That means, setting this flag to `NO_COLLATION` during runtime could lead to some unexpected behaviors/errors. Besides, the main issue in [this issue](https://github.com/react-native-community/async-storage/issues/178) is `database is locked` error - which is also addressed by this PR.

File: android/src/main/java/com/reactnativecommunity/asyncstorage/ReactDatabaseSupplier.java [hunk 3]
```diff
@@ -7,15 +7,14 @@
...
-import javax.annotation.Nullable;
-
 import android.content.Context;
 import android.database.sqlite.SQLiteDatabase;
 import android.database.sqlite.SQLiteException;
 import android.database.sqlite.SQLiteOpenHelper;
-
 import com.facebook.common.logging.FLog;
 import com.facebook.react.common.ReactConstants;
+import java.io.File;
```
Comment (tido64) [inline android/src/main/java/com/reactnativecommunity/asyncstorage/ReactDatabaseSupplier.java:16]: Unused?

**Review by tido64 — APPROVED** [review]

**Comment by tido64** [comment]
> According to [Android docs, one needs to be consistent with what collation rules were used when creating the database](https://developer.android.com/reference/kotlin/android/database/sqlite/SQLiteDatabase#no_localized_collators). That means, setting this flag to `NO_COLLATION` during runtime could lead to some unexpected behaviors/errors. Besides, the main issue in [this issue](https://github.com/react-native-community/async-storage/issues/178) is `database is locked` error - which is also addressed by this PR.

Does this mean that the problem is that we tried to open a database that was already opened in a different locale? Wouldn't restarting the app already close the database? Or is the implementation such that it is possible to leak connections?

**Comment by Krizzu** [comment]
> Does this mean that the problem is that we tried to open a database that was already opened in a different locale? 

I believe this is the case we see here.

> Wouldn't restarting the app already close the database? Or is the implementation such that it is possible to leak connections?

This is exactly what I think is happening - changing locale restarts the activity, not the app process, so it leaks the db connection.

**Comment by tido64** [comment]
> This is exactly what I think is happening - changing locale restarts the activity, not the app process, so it leaks the db connection.

The repro steps in #178 mentions killing the app though. I thought that would close the connection?

**Comment by Krizzu** [comment]
> The repro steps in #178 mentions killing the app though. I thought that would close the connection?

This is the most confusing part of the whole issue. I also thought the connection will be closed, but apparently it's not 🤷‍♂️ Dunno if we're not missing something here.

**Comment by Krizzu** [comment]
@tido64 I'll go ahead and merge this. Thanks for reviewing 🙏

**Pull request merged**